### ISSUE TITLE
ValidationException for get-only property with 3.0-RC01

### ISSUE THREAD

**Issue by StefanBertels** [JoshClose/CsvHelper#747]
After updating to RC01 I get some ValidationException for public get-only properties when reading data:

```
class MyRecord
{
   public int MyValue {get; set;}
   public bool IsNegative => MyValue < 0;
}
```

Error: `CsvHelper.ValidationException: Header 'IsNegative' was not found.`
Worked with 3.0.0-chi06. No custom mapping involved.

I could workaround this by making it `internal` but I think it should be possible to _write_ csv files with get-only properties and just ignore these properties when _reading_.

**Comment by JoshClose** [comment]
Did you miss the rest of the exception message? `Header 'IsNegative' was not found. If you are expecting some headers to be missing and want to turn this validation check off, set the configuration ThrowOnBadHeader to false.`

You just need to do `csv.Configuration.ThrowOnBadHeader = false;`. This was turned on by default so that people would know if their class isn't matching up with their file.

**Comment by JoshClose** [comment]
If you don't think this should be on by default, I'd like to hear your argument. I originally didn't want it on, but I do think it's helpful to know if something is wrong. It only validates against mapped properties, so if you have a mapping, it won't be an issue. It really only becomes a problem when auto mapping, but I still think it's good to know, and you can turn it off if you want.

Also, there is a new factory for building maps without having to use another class.

```cs
var map = new Factory().CreateClassMapBuilder<MyRecord>()
	.Map( m => m.MyValue )
	.Build();
csv.Configuration.RegisterClassMap( map );
var records = csv.GetRecords<MyRecord>().ToList();
```

**Comment by StefanBertels** [comment]
ThrowOnBadHeader probably might suppress other errors, too.

I understand the "format validation" reason. But I think properties without `set` never can be written to, so automapping should ignore them by default when reading. I guess you always have one map for read and write. But reading and writing does not need to be symmetrical in general (see custom mapping). 

Asymmetrical automapping is a quite clear convention (code first style): include all props. If there is a (public) `set` you can read this column from file. If there is a (public) `get` you can write this to file. Works without exceptions. 

Point of view: there are no public properties, just properties with public setter and/or public getter.

**Comment by JoshClose** [comment]
The auto mapper will map all properties because the map is used for reading and writing. Having no setter but a getter means it can be used for writing. I think the `ValidateHeader` method needs to do some checks like the reader does. There is a `CanRead` method that is checked when reading, and if `false` is returned, the property is ignored. I'll have it call this also, and those properties will be ignored when validating. I'll see if I can find other scenarios where they should be ignored also.

**Comment by JoshClose** [comment]
My goal for this library is to have things work really easily out of the box without any configuration. I want the scenario of auto mapping to just work. This `CanRead` change fixes that in most cases. The one I struggle with though is a class having extra writable properties and someone using that. Maybe this is ok though.

**Comment by StefanBertels** [comment]
Sorry, what does "extra writable property" mean?

**Comment by JoshClose** [comment]
If the class has 5 properties that all have public setters, but the CSV file only has 3 columns. There are 2 extra properties that can be written to (have public setters). Bad wording. I don't think this will be the case often though.

**Comment by StefanBertels** [comment]
I understand. Probably there are different opinions. But maybe it's cleaner to use one rule:
every public get property is written to file and every public set property is read from file. 
This might result in additional errors (we don't like that) but this is cleaner / code-first.
But there should be an additional switch to allow extra set properties. The class maybe can't be changed.

Maybe custom mapping is a valid recommendation for such a case.

Another reason to throw error for this: even if those 2 extra properties have public setter by intend and are never in any file (might be filled later), you probably want an error if one of the other 3 is missing in _one_ file.

**Comment by StefanBertels** [comment]
But: isn't this current behaviour? (willthrowonmissingfields)

**Comment by JoshClose** [comment]
`ThrowOnMissingField` will throw on any `GetField( string name )` or `GetField( int index )` where the field doesn't exist. If you're using `GetRecord[s]` the header validation will have caught that already. If you do `GetRecords` and the file is blank or has a header that is garbage and `ThrowOnMissingField` doesn't get thrown because it never actually tried to read a field.

These 2 seem very similar. Maybe `ThrowOnBadHeader` can be removed, and the `ValidateHeader` method check can be based on `ThrowOnMissingField` instead, and it could throw a `MissingFieldException` instead of a `ValidationException`.

Would that make more sense?

**Comment by StefanBertels** [comment]
1. Sorry for confusion, in my comment above I meant ThrowOnBadHeader because we're talking about validation header against classmap.

2. I think there should be two different exceptions. It's another problem whether the field is missing (i.e. there are not enough columns in that data row) or whether the header row misses a column with some specific name. I don't know whether you want fewer options / reduce functionality (clean up) but there are probably use cases to suppress both errors individually.

If we agree that default is somewhat strict so there will be exceptions (MissingFieldValue) thrown if columns are missing in some data row _and_ there will be exceptions (MissingColumnInHeader) thrown if the class map has a mapping without some matching column (name) in actual file's header.

You should still have two config options (ThrowOnMissingFieldValue, ThrowOnMissingColumnHeader). I changed `ThrowOnMissingField` and `ThrowOnBadHeader` names here to make it more clear, but you probably have your own idea of names (and being compatible).

Alternatively you could replace the switch (true/false) by some function. OnMissingFieldValue could be a function that returns custom default:

```
// some reader config property (set to something different, if you do not want to throw)
Func<ReadingContext, int, string> OnMissingFieldValue = (readingContext, colIndex) =>
{
	if (readingContext.ReaderConfiguration.IgnoreBlankLines)
	{
	    throw new MissingFieldException(readingContext, $"Field at index '{colIndex}' does not exist.");
	}

	return default(string);
};


public virtual string GetField( int index )
{
	CheckHasBeenRead();

	// Set the current index being used so we
	// have more information if an error occurs
	// when reading records.
	context.CurrentIndex = index;

	if( index >= context.Record.Length )
	{
		return OnMissingFieldValue( context, index );
	}

	var field = context.Record[index];

	return field;
}
```

This way you could suppress exceptions with:
`myconfig.OnMissingFieldValue = (readingContext, colIndex) => return default(string);`

I'm not sure about `readingContext.ReaderConfiguration.IgnoreBlankLines` condition (is current code correct?) so I just left it there.

Anyway: this type of handlers makes it easier to customize (more flexibel, clean) but some users of your library might find it easier to have just some bool property to switch such behaviour. But you already went this way, see e.g. PrepareHeaderForMatch (which I like).

**Comment by JoshClose** [comment]
Yeah... it's not very intuitive to have a `OnMissingField` config and they need to do `OnMissingField = null;` or something to ignore. I actually do like that idea better. I like giving the user the ability to do whatever they want, instead of having a bunch of little configurations. I may go this route.

**Comment by JoshClose** [comment]
I removed several flags in favor of callback methods. I put a description message in the exceptions thrown so the user will know how to turn them off if need be.

I used the suffix `Callback` instead of the prefix `On` since that has a special meaning with events in .NET.

13dbb61caa7dda5c1172287260c6e94ee15a9723

**Comment by StefanBertels** [comment]
Hi, I just had a problem with RC05: "Header matching ... not found". Seems Map(_ => _.MyColumn).ConvertUsing(...) is the problem, csv field name is different to MyColumn. I'm just trying to build a test case.

**Comment by JoshClose** [comment]
I just released 3.0.0. I'll try and get to this soon.
